### ISSUE TITLE
Conceptual discussion on lecture 6

### ISSUE THREAD

**Issue by ekuiter** [SoftVarE-Group/Course-on-Software-Product-Lines#32]
I am skeptical about the concept for lecture 6 (components, services, and frameworks). For me, this is the hardest lecture to give, and also the most confusing exercise class. (This is partially due to the exercise tasks, but also due to the corresponding lecture, which this issue is about.)

- The lecture does not really introduce anything new. Components are an architecture-level instantiation of design patterns already earlier discussed in the lecture. In the exercise, I regularly struggle to delineate this clearly. Each of the implementation lectures clearly introduces a novel concept (2-regular language features, 3-cloning, 5-language-independent tools, 7-new language features), except for this lecture IMO..
- What is the conceptual difference between components and (micro-)services? Yes, by moving components into network environments, some small things change. Yes, microservices are typically tied to a particular organization structure/philosophy. But these feel like nuances which can be explained on a single slide, not treated as "big" technical implementation techniques.
- The lecture explains in great detail what components, services, and frameworks are. However, it is barely explained how this is connected to variability and SPLs. The lego images are unfortunately handwavy - I am not able to clearly explain how to implement an SPL based on components or services. I also do not know of any examples of such SPLs. Both concepts are barely mapped onto concrete code, in contrast to the other lectures.
- What is the conceptual difference between components and frameworks? As far as I can tell, only preplanning and inversion of control. How does this justify that we need to write glue code for components and not for frameworks? A very regular use case of frameworks is the adaptation of existing component code - so glue code is necessary in that case. Vice versa, there are also standardized components, which do not require writing glue code at all to switch out. In general, I believe there is no clearcut distinction between components and frameworks, which makes it really hard to explain in the exercise.

IMO, large parts of this lecture can be dropped without losing anything important. Of the three blocks, frameworks seem the least confusing (although they are still just applications of things discussed earlier in the lecture). (see #18)

Alternatively, we could improve the definitions of these concepts and give more concrete examples. To some degree, this could also be solved by moving lecture 6 into the ad hoc part of the lecture series, as I would not associate any of these techniques with the classical definition of the term SPL. (see #20)

**Comment by ekuiter** [comment]
*tl;dr: After rereading my comments above, this is basically the same criticism. We should discuss at some point.*

We just discussed the exercise sheet 6. Here's a recap of my points of confusion, which impact both the exercise and the lecture.

**Components**

- What is the use case of components for SPLs?
    - If I implement my components myself, I am already likely to implement them in such a way that they have good interfaces, which means I will probably end up with plugins anyway.
    - If I use third party components (which I think is what we aim at, using others code, with the whole market ecosystem idea), I need to "glue" them together as explained in the lecture. But this is just what we do in normal software design, where is the product line aspect?
    - If I use third party components *so I can exchange them freely, and realize choices in a product line,* that practically means to write an adapter, bridge or facade to the third party component, which will create a standardized interface. So, we basically end up again with applying design patterns and plugins.
- What is the difference between a component, library, and module?
- Do we have an example for a component based product line? If not, why include this in the lecture?

**Services**

- What is the difference between a service and a component? If it's only the distributed and organizational aspects, how does this affect SPL engineering at all?
- Do we have an example for a (micro)service based product line? If not, why include this in the lecture?

Overall, to me components and (micro)services mostly sound like "buzzwords" in the context of this lecture. They are definitely interesting to discuss in general, but I only see very superficial ties to product line engineering. They rather seem like variations on techniques which we've discussed before. Hence our troubles in the exercise.
